### PR TITLE
tests: Use libvmi instead of 'NewRandomVMIWithEphemeralDiskAndConfigDriveUserdata

### PR DESCRIPTION
**Pull request by ormergi**
### What this PR does
Use libvmi directly at the caller instead of `NewRandomVMIWithEphemeralDiskAndConfigDriveUserdata`.

Before this PR:

After this PR:

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/utils.go
Comment (fossedihelm) [inline tests/utils.go:535]: please, use `libvmi.WithResourceMemory("512M")`
or
directly address `NewRandomVMIWithEphemeralDiskAndConfigDriveUserdataHighMemory` in this PR (just one usage)

**Review by fossedihelm — APPROVED** [review]
Thanks! Added a suggestion. Feel free to ignore it with the promise of a follow-up to cover `NewRandomVMIWithEphemeralDiskAndConfigDriveUserdataHighMemory` :)
Thank you

File: tests/utils.go
Comment (ormergi) [reply tests/utils.go:535]: I will go with option 1 because we plan to remove `NewRandomVMIWithEphemeralDiskAndConfigDriveUserdataHighMemory` as well.
Thanks for looking at this :slightly_smiling_face:

**Review by fossedihelm — APPROVED** [review]
Thank you!

File: tests/vmi_cloudinit_test.go
Comment (orelmisan) [inline tests/vmi_cloudinit_test.go:178]: This test is about the base64 encoded user data in cloudInitConfigDrive:
```suggestion
					libvmi.WithCloudInitConfigDriveEncodedUserData(userData),
```
This option function does not exist and should be added.

File: tests/utils.go
Comment (orelmisan) [inline tests/utils.go:531]: The `containerImage` parameter is no longer used.
This could be handled in a follow-up.
WDYT?

**Review by orelmisan — CHANGES_REQUESTED** [review]
Thank you for the cleanup @ormergi.

File: tests/utils.go
Comment (fossedihelm) [reply tests/utils.go:531]: IMHO the follow up should be Remove the deprecated function :)

File: tests/vmi_cloudinit_test.go
Comment (fossedihelm) [reply tests/vmi_cloudinit_test.go:178]: The test is not about the encoded base64, but the non-encoded one.
Also, the test previously was already using non-encoded data:
```
AddCloudInitConfigDriveData(vmi, "disk1", userData, "", false) //false means non-encoded
```

Comment (orelmisan) [reply tests/vmi_cloudinit_test.go:178]: Either the test or the title is misleading.

Comment (fossedihelm) [reply tests/vmi_cloudinit_test.go:178]: Got you! Thanks! Looking at the history of the test, it looks like it was created with no encoding, so this makes me think that it is the test description that is misleading.

Comment (ormergi) [reply tests/vmi_cloudinit_test.go:178]: My guess is the test author copy-pasted the CloudInitNoCloud tests but forgot to remove the 'base64' part from description ([PR](https://github.com/kubevirt/kubevirt/pull/2120)).

In terms of coverage we are good because there are [unit tests](https://github.com/kubevirt/kubevirt/blob/main/pkg/cloud-init/cloud-init_test.go#L464) that covers both options (use-data and base64-user-data).

I think we should rename the test description to not mention base64 as history shows the test verifies regular user-data and we have base64 logic covered in unit-tests.
@orelmisan @fossedihelm  WDYT?

File: tests/utils.go
Comment (ormergi) [reply tests/utils.go:531]: We do plan to remove this function, I recall @orenc1 is on it.

File: tests/vmi_cloudinit_test.go
Comment (orelmisan) [reply tests/vmi_cloudinit_test.go:178]: AFAIU, `[test_id:3180] with cloudInitConfigDrive userData source"` is testing it.
So it makes sense to me to test the base64 logic.

File: tests/utils.go
Comment (orenc1) [reply tests/utils.go:531]: opened #11327

File: tests/vmi_cloudinit_test.go
Comment (fossedihelm) [reply tests/vmi_cloudinit_test.go:178]: I'm ok with both of the decisions:
- change the test description
- test w/ the encoded data, adding a new `Option` (in a separate commit please).

However, the test duplication @orelmisan discovered, deserve a punctual attention and maybe a revision of the entire cloudinit e2e test files.

Comment (ormergi) [reply tests/vmi_cloudinit_test.go]: Done, test is fixed to verify encoded user data.

On a followup PR we can revisit this whole test suite and consolidate/remove tests.

**Comment by ormergi** [comment]
@orelmisan @fossedihelm PTAL :slightly_smiling_face:

File: tests/libvmi/cloudinit.go
Comment (dankenigsberg) [inline tests/libvmi/cloudinit.go:81]: How could this pass? You are providing a non-encoded data here.

(BTW I don't understand the point of the encoded field, and would vote to stop testing it in e2e)

Comment (ormergi) [reply tests/libvmi/cloudinit.go]: My bad I pushed the wrong changes, PTAL

**Review by fossedihelm — APPROVED** [review]

**Review by enp0s3 — COMMENTED** [review]
/approve

**Comment by brianmcarey** [comment]
/hold 

@ormergi - it looks like there are two tests in the compute lanes that are failing consistently on this PR. Placing a hold just to stop needless retests.

File: tests/libvmi/cloudinit.go
Comment (fossedihelm) [inline tests/libvmi/cloudinit.go:89]: I don't understand what this function does.
The returned value is not used and the volume is not added.
Maybe something is missing here? :)

**Comment by ormergi** [comment]
rebased

File: tests/libvmi/cloudinit.go
Comment (ormergi) [reply tests/libvmi/cloudinit.go]: Please have another look :slightly_smiling_face:

**Review by fossedihelm — COMMENTED** [review]
Thank you @ormergi 
Some comments below

File: pkg/libvmi/cloudinit.go
Comment (ormergi) [reply pkg/libvmi/cloudinit.go:118]: The problem with introducing new Cirros factory outside `libfact` is that we will need to expose `cirrosMemory()`, which I recall Edy was against. 

The easiest way to get out of it is using Alpine VMs instead because they dont need the initial cloud init configuration (w/o it, boot take long time).
The resource footprint should be similar.

Comment (ormergi) [reply pkg/libvmi/cloudinit.go:118]: Also, I recall @dankenigsberg suggested adding no-cloud-init option to libvmi to avoid this cloud-init volume mess.
Then it should look like so:
```
    libfact.NewCirros(
        libvmi.NoCloudInitNoCloud(),
        libvmi. WithCloudInitConfigDriveUserData()
    ...
```

WDYT?

Comment (fossedihelm) [reply pkg/libvmi/cloudinit.go:118]: What is the meaning of `NoCloudInitNoCloud`?
I think that it will be great to have the default cloudinit in cirros only if no other cloudinit is specified in the options.
In this specific case, if we want to move this forward , we can use alpine

Comment (ormergi) [reply pkg/libvmi/cloudinit.go:118]: The idea is to have `NoCloudInitNoCloud` remove existing cloudInitNoCloud volume, using it combined with `WithCloudInitConfigDriveUserData` will avoid the conflict

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [inline pkg/libvmi/cloudinit.go:104]: I don't understand the bug that this commit tries to fix, but I don't believe that this is an acceptable semantics: you should not silently keep the former volume.

File: tests/vmi_cloudinit_test.go
Comment (ormergi) [reply tests/vmi_cloudinit_test.go]: My bad, fixed.

File: pkg/libvmi/cloudinit.go
Comment (ormergi) [reply pkg/libvmi/cloudinit.go]: Unfortunately its not possible to use alpine VMs because it works the same as cirros in a sense that `NewAlpine` factory adds cloud-init user-data volume by default to prevent long boot time :disappointed: 

Fallback to adding without-cloud-init-no-cloud `libvmi` option combined with using cloud-init config-drive `libvmi` options.
Please take another look :slightly_smiling_face:

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [inline pkg/libvmi/cloudinit.go:103]: It seems that you add a function into `libvmi` just in order to fix a limitation in a higher-level `libvmifact`. This sounds a bit awkward, maybe I don't understand you.

File: tests/vmi_cloudinit_test.go
Comment (fossedihelm) [inline tests/vmi_cloudinit_test.go:642]: I think we should be more reliable and should remove every cloudinitnocloud volume, without breaking the cycle, even if there could be only one.
In general, I am still non totally convinced about this, but for the moment it could be the better tradeoff.
I'm ready to lgtm once we are sure we are removing all the cloudinitnocloud volumes. (hint: I think we should create a new volume list to be assigned at the end of the for cycle).
If you think that the current implementation is better, I am ready to be convinced :)
Thank you!

Comment (ormergi) [reply tests/vmi_cloudinit_test.go:642]: Thanks for the feedback and reviews, its is much appreciated :slightly_smiling_face: 

There were some discussions offline about having a factory for cirros w/o cloud-init as an alternative, where it create a vmi with no cloudinit columes and the relevant cloudinit volume is added.
Instead of creating a vmi with some cloudinit volume, remove it, and add the relevant cloudinit.

There is a PR that implements it and remove  `NewRandomVMIWithEphemeralDiskAndConfigDriveUserdata` #11923.

**Pull request closed without merging**
